README.md
# Joulie-16-LCD
buildroot config for a Battery info screen with web ui for the Raspberry Pi
reworked to utilise the UDP Multicast protocoll enabled via DIP Switch 8 since version 1.14 of the JouLie-16 

## Required HW:

× Raspberry Pi

× 20x4 I2C LCD

× 3.3V to 5V Level shifter

× Class 10 Micro SD Card with at least 4GB

×Jumper wires and/or a small circuit board


## Instructions for building:

* Get the latest stable Buildroot release from (https://buildroot.org/downloads/buildroot-2019.08.tar.gz)
* Unpack the buildroot tar
* Clone this repo
* Move the overlay dir into the unpacked buildroot dir
* Move the config file into the buildroot dir and rename it to .config
* Open a Terminal in the buildroot dir
* Run `make menuconfig` and check the `Target options` to make sure the config was loaded
  - Target Architecture (ARM (little endian))
  - Target Binary Format (ELF)
  - Target Architecture Variant (arm1176jzf-s)
* Quit the menuconfig by hitting *ESC* until you return to the Terminal
* Run `make` to start building your image. This requires an active internet connection to download all the packages. This will take a couple of hours depending on your system.
* Once finished, navigate your file browser to the **output/images** dir where you will find a **sdcard.img** file. This is your SD card image.
* **Alternatively you can download a prebuilt Image here: (https://drive.google.com/drive/folders/1bVy2NC1GEXSF4w1Ivt7RKugK-Z9H2YX_?usp=sharing)**
* Write that image file to your SD card either using dd on Linux systems or using https://etcher.download
* Unmount the SD Card and plug it into your Pi
* Connect the I2C LCD to the Pi according to the instructions below. Most LCD's will run off of 3.3V with usable but not good contrast, so a level shifter is recommended.

Connect the components as follows. You can either use jumper wires or solder the components to a piece of through-hole board.

LCD SCL → Shift 1 → Pi Pin 5

LCD SDA → Shift 2 → Pi Pin 3

LCD GND → Shift GND → Pi Pin 6

LCD Vcc → Pi Pin 2

Shift HV → Pi Pin 4

Shift LV → Pi Pin 1

* When directly connecting the LCD to the Pi, power the LCD from Pin 1 instead of Pin 2 from the Pi
* Connect the Pi to the same network as the BMS. This can be through a switch as long as they're in the same network and the Pi can receive the UDP Multicast packets
* Power up the Pi
* The Pi should boot within ~2 minutes
* You should now see the battery status on the LCD
* Now you can open up a browser and go to (http://autarctech-bms/) to view the web interface.
* If you are directly connected to the BMS then call up this address: http://10.214.161.1
* Debugging via SSH:
  - Username: root
  - Password: AutarcTech

### To Do:
- [ ] Proper interactive Web Interface
- [ ] CANBUS Mode
- [ ] Cloud?
- [ ] Better local screen. Maybe a TFT? Or a HDMI Screen?


CHANGELOG
———————2.3.2———————
  · Configured systemd-network to be both DHCP Client and Server for use in existing networks and independent setups
  · ditched queue in favor of deque
    → allows for old entrys to fall off the queue instead of blocking when queue is full
    → Now all threads just peek at the latest value in the queue and no thread actually pop's from the queue
  · Added custom battery characters for the display
  · Added scaling factor for the current measurement. To use simply place a "currentscale.txt" into the SD-card's boot partition and put a scaling factor in the first line.
  · Started implementing multi-bms compatibility
  · Improved web-ui and added german language


overlay/scripts/LCD_Meter.py
import I2C_LCD_driver
from time import sleep
import serial
import random
import sys

mylcd = I2C_LCD_driver.lcd()
for i in range(15):
	try:
		ser = serial.Serial('/dev/ttyACM0', timeout=1)
		sleep(0.25)
	except:
		print("Error trying to connect")

try:
	ser.is_open
except:
	print("Serial port could not be opened.")
	mylcd.lcd_display_string("Serial Error!",1)
	sys.exit()


counter_write = 0
SOC = 0
Voltage = 52447
Current = 100
Power = Voltage * Current
try:
	f = open("energy.txt", "r")
	read = f.readline()
	if float(read) > 0:
		Energy = float(read)
	f.close()
except:
	Energy = 0
direction = 'pos'

mylcd.lcd_display_string("Battery SOC:",1)
try:
	while True:
		try:
			data = ser.readline()
			data = str(data).split(";")
			Voltage = int(data[36])
			SOC = int(data[35])/1000
			Current = int(data[33])
			Power = (Voltage/1000.0) * (Current/1000.0)
			if Power > 0:
				Energy += Power		# Energy in Ws
			kilowatthours = Energy / (3600000*4)
			mylcd.lcd_display_string((str(round(SOC, 1))+"%").rjust(5),2)
			
			mylcd.lcd_display_string((str(round(Voltage/1000.0,2))+"V").rjust(6), 2, 14)
			mylcd.lcd_display_string((str(round(Current/1000.0,1))+"A").ljust(7),3)
			mylcd.lcd_display_string((str(round(Power,2))+"W").rjust(9),3 ,11)
			mylcd.lcd_display_string("",4)
			mylcd.lcd_write_char(219)
			if Current > 4000:
				mylcd.lcd_write_char(127)
			elif Current < -4000:
				mylcd.lcd_write_char(126)
			else:
				mylcd.lcd_write_char(165)
			mylcd.lcd_display_string((str(round(kilowatthours,1))+"kWh").rjust(18),4,2)
			counter_write += 1
			if counter_write >= 100:
				f = open("energy.txt", "w+")
				f.write(str(Energy))
				f.close()
				counter_write = 0
			ser.reset_input_buffer()
			sleep(0.25)
		except (KeyboardInterrupt, SystemExit):
			break
		except:
			pass
except (KeyboardInterrupt, SystemExit):
		print("Exiting…")
		ser.close()


overlay/scripts/I2C_LCD_driver.py
# -*- coding: utf-8 -*-
# Original code found at:
# https://gist.github.com/DenisFromHR/cc863375a6e19dce359d

"""
Compiled, mashed and generally mutilated 2014-2015 by Denis Pleic
Made available under GNU GENERAL PUBLIC LICENSE

# Modified Python I2C library for Raspberry Pi
# as found on http://www.recantha.co.uk/blog/?p=4849
# Joined existing 'i2c_lib.py' and 'lcddriver.py' into a single library
# added bits and pieces from various sources
# By DenisFromHR (Denis Pleic)
# 2015-02-10, ver 0.1

"""

# i2c bus (0 -- original Pi, 1 -- Rev 2 Pi)
I2CBUS = 1

# LCD Address
ADDRESS = 0x27

import smbus
from time import sleep

class i2c_device:
   def __init__(self, addr, port=I2CBUS):
      self.addr = addr
      self.bus = smbus.SMBus(port)

# Write a single command
   def write_cmd(self, cmd):
      self.bus.write_byte(self.addr, cmd)
      sleep(0.0001)

# Write a command and argument
   def write_cmd_arg(self, cmd, data):
      self.bus.write_byte_data(self.addr, cmd, data)
      sleep(0.0001)

# Write a block of data
   def write_block_data(self, cmd, data):
      self.bus.write_block_data(self.addr, cmd, data)
      sleep(0.0001)

# Read a single byte
   def read(self):
      return self.bus.read_byte(self.addr)

# Read
   def read_data(self, cmd):
      return self.bus.read_byte_data(self.addr, cmd)

# Read a block of data
   def read_block_data(self, cmd):
      return self.bus.read_block_data(self.addr, cmd)


# commands
LCD_CLEARDISPLAY = 0x01
LCD_RETURNHOME = 0x02
LCD_ENTRYMODESET = 0x04
LCD_DISPLAYCONTROL = 0x08
LCD_CURSORSHIFT = 0x10
LCD_FUNCTIONSET = 0x20
LCD_SETCGRAMADDR = 0x40
LCD_SETDDRAMADDR = 0x80

# flags for display entry mode
LCD_ENTRYRIGHT = 0x00
LCD_ENTRYLEFT = 0x02
LCD_ENTRYSHIFTINCREMENT = 0x01
LCD_ENTRYSHIFTDECREMENT = 0x00

# flags for display on/off control
LCD_DISPLAYON = 0x04
LCD_DISPLAYOFF = 0x00
LCD_CURSORON = 0x02
LCD_CURSOROFF = 0x00
LCD_BLINKON = 0x01
LCD_BLINKOFF = 0x00

# flags for display/cursor shift
LCD_DISPLAYMOVE = 0x08
LCD_CURSORMOVE = 0x00
LCD_MOVERIGHT = 0x04
LCD_MOVELEFT = 0x00

# flags for function set
LCD_8BITMODE = 0x10
LCD_4BITMODE = 0x00
LCD_2LINE = 0x08
LCD_1LINE = 0x00
LCD_5x10DOTS = 0x04
LCD_5x8DOTS = 0x00

# flags for backlight control
LCD_BACKLIGHT = 0x08
LCD_NOBACKLIGHT = 0x00

En = 0b00000100 # Enable bit
Rw = 0b00000010 # Read/Write bit
Rs = 0b00000001 # Register select bit

class lcd:
   #initializes objects and lcd
   def __init__(self):
      self.lcd_device = i2c_device(ADDRESS)

      self.lcd_write(0x03)
      self.lcd_write(0x03)
      self.lcd_write(0x03)
      self.lcd_write(0x02)

      self.lcd_write(LCD_FUNCTIONSET | LCD_2LINE | LCD_5x8DOTS | LCD_4BITMODE)
      self.lcd_write(LCD_DISPLAYCONTROL | LCD_DISPLAYON)
      self.lcd_write(LCD_CLEARDISPLAY)
      self.lcd_write(LCD_ENTRYMODESET | LCD_ENTRYLEFT)
      #sleep(0.2)


   # clocks EN to latch command
   def lcd_strobe(self, data):
      self.lcd_device.write_cmd(data | En | LCD_BACKLIGHT)
      #sleep(.0005)
      self.lcd_device.write_cmd(((data & ~En) | LCD_BACKLIGHT))
      sleep(.0001)

   def lcd_write_four_bits(self, data):
      self.lcd_device.write_cmd(data | LCD_BACKLIGHT)
      self.lcd_strobe(data)

   # write a command to lcd
   def lcd_write(self, cmd, mode=0):
      self.lcd_write_four_bits(mode | (cmd & 0xF0))
      self.lcd_write_four_bits(mode | ((cmd << 4) & 0xF0))

   # write a character to lcd (or character rom) 0x09: backlight | RS=DR<
   # works!
   def lcd_write_char(self, charvalue, mode=1):
      self.lcd_write_four_bits(mode | (charvalue & 0xF0))
      self.lcd_write_four_bits(mode | ((charvalue << 4) & 0xF0))
  
   # put string function with optional char positioning
   def lcd_display_string(self, string, line=1, pos=0):
    if line == 1:
      pos_new = pos
    elif line == 2:
      pos_new = 0x40 + pos
    elif line == 3:
      pos_new = 0x14 + pos
    elif line == 4:
      pos_new = 0x54 + pos

    self.lcd_write(0x80 + pos_new)

    for char in string:
      self.lcd_write(ord(char), Rs)

   # clear lcd and set to home
   def lcd_clear(self):
      self.lcd_write(LCD_CLEARDISPLAY)
      self.lcd_write(LCD_RETURNHOME)

   # define backlight on/off (lcd.backlight(1); off= lcd.backlight(0)
   def backlight(self, state): # for state, 1 = on, 0 = off
      if state == 1:
         self.lcd_device.write_cmd(LCD_BACKLIGHT)
      elif state == 0:
         self.lcd_device.write_cmd(LCD_NOBACKLIGHT)

   # add custom characters (0 - 7)
   def lcd_load_custom_chars(self, fontdata):
      self.lcd_write(0x40);
      for char in fontdata:
         for line in char:
            self.lcd_write_char(line)         
         




overlay/scripts/server.py
#!/usr/bin/env python3

__author__ = "Elias Ibis"
__copyright__ = "Copyright 2019, AutarcTech GmbH"
__license__ = "GPL"
__version__ = "2.3.3"
__maintainer__ = "Elias Ibis"
__email__ = "elias.ibis@autarctech.de"
__status__ = "Development"

#Loading all required modules
import I2C_LCD_driver
import subprocess
import json
import sys
from threading import Thread
import collections
from bottle import run, post, request, response, get, route, static_file, HTTPResponse, error
from time import sleep
import random
import socket
import crcmod
import os

def map(x, in_min, in_max, out_min, out_max):
    return max(min(out_max, int((x-in_min) * (out_max-out_min) / (in_max-in_min) + out_min)), out_min)



try:
	with open('/boot/currentscale.txt') as f:
		current_scaling_factor = float(f.readline())
except:
	current_scaling_factor = 1
	pass
customchars = [
	[0x0E,0x1F,0x11,0x11,0x11,0x13,0x17,0x1F],
	[0x0E,0x1F,0x11,0x11,0x13,0x17,0x1F,0x1F],
	[0x0E,0x1F,0x11,0x13,0x17,0x1F,0x1F,0x1F],
	[0x0E,0x1F,0x13,0x17,0x1F,0x1F,0x1F,0x1F],
	[0x0E,0x1F,0x1F,0x1F,0x1F,0x1F,0x1F,0x1F],
	[0x04,0x0E,0x1F,0x04,0x04,0x04,0x04,0x04],
	[0x04,0x04,0x04,0x04,0x04,0x1F,0x0E,0x04]
]


#######UDP Multicast Code

crc16 = crcmod.mkCrcFun(0x18005, rev=True, initCrc=0x0000, xorOut=0x0000)                           # CRC Setup for Joulie-16

ANY = "0.0.0.0" 
MCAST_ADDR = "239.255.73.100"
MCAST_PORT = 1500
MCAST_TTL = 2
BMS_LIST = []

try:
	# Create a UDP socket
	sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)

	# Allow multiple sockets to use the same PORT number
	sock.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR,1)

	# Bind to the port that we know will receive multicast data
	sock.bind((ANY,MCAST_PORT))

	# Tell the kernel that we are a multicast socket
	sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, MCAST_TTL)

	# Tell the kernel that we want to add ourselves to a multicast group
	# The address for the multicast group is the third param


	status = sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, socket.inet_aton(MCAST_ADDR) + socket.inet_aton(ANY))

	# setblocking(0) is equiv to settimeout(0.0) which means we poll the socket.
	# But this will raise an error if recv() or send() can't immediately find or send data. 
	sock.setblocking(1)
	sock.settimeout(1)
	count = 0
except:
	print("An error has okuu'd")
	sys.exit()

#######End UDP Multicast Code



#initialize queue for multithreading
q = []
#q.append(collections.deque(maxlen=2)

#Initializing I2C LCD on address defined by I2C_LCD_Driver module
try:
	mylcd = I2C_LCD_driver.lcd()
except:
	pass
try:
	mylcd.lcd_load_custom_chars(customchars)
	mylcd.lcd_display_string(__version__,1)
except:
	pass


#Thread that listens to the network for new data and tries to put it into the queue
def scavenge_data_q(q):
	while True:
		try:
			data, addr = sock.recvfrom(1024)
			data = str(data).split(",")
			if data[2] == '100':
				if data[1] not in BMS_LIST:
					BMS_LIST.append(data[1])
					q.append(collections.deque(maxlen=2))
				try:
					q[BMS_LIST.index(data[1])].append(data)
				except:
					sleep(0.1)
					pass
		except KeyboardInterrupt:
			sys.exit()
		except:
			print("No data received. Is the BMS connected to the network and turned on?")
		sleep(0.1)


#Thread that updates the I2C LCD Screen every second. Deque peek
def update_lcd():
	try:
		sleep(1)
		mylcd.lcd_display_string("Battery Status:",1)
		while True:
			for i in range(len(BMS_LIST)):
				for n in range(10):
					data = q[i][0]
					Voltage = float(data[-8])
					SOC = float(data[-6])
					Current = float(data[-9]) * current_scaling_factor
					Power = Voltage * Current
					mylcd.lcd_display_string((str(round(SOC, 1))+"%").rjust(5),2)
					mylcd.lcd_display_string((str(round(Voltage,2))+"V").rjust(6), 2, 14)
					mylcd.lcd_display_string((str(round(Current,1))+"A").ljust(7),3)
					mylcd.lcd_display_string((str(round(Power,2))+"W").rjust(9),3 ,11)
					#mylcd.lcd_display_string(ni.ifaddresses('wlan0')[ni.AF_INET][0]['addr'].rjust(18),4, 2)
					mylcd.lcd_display_string(("ID: " + str(BMS_LIST[i])).rjust(18),4, 2)
					mylcd.lcd_display_string("",4)
					mylcd.lcd_write_char(map(int(SOC),20,80,0,4))
					if Current > 2:
						mylcd.lcd_write_char(5)
					elif Current < -2:
						mylcd.lcd_write_char(6)
					else:
						mylcd.lcd_write_char(165)
					
					sleep(1)
	except KeyboardInterrupt:
		sys.exit()
	except:
		pass


#This function is parsing the raw BMS data into a JSON for usage with javascript. Deque peek
def get_json():
	datadict = {'BMSCount':len(BMS_LIST)}
	for i in range(len(BMS_LIST)):
		data = q[i][0]
		Voltage = float(data[-8])
		SOC = float(data[-6])
		Current = float(data[-9]) * current_scaling_factor
		Power = Voltage * Current
		datadict.update({"BMS" + str(i):{
		"SOC":{"Unit":"%","Value":SOC},
		"Voltage":{"Unit":"V","Value":Voltage},
		"Current":{"Unit":"A","Value":Current},
		"Power":{"Unit":"W","Value":Power}
		,"SW":{"RawData":data}}})
		json_out = json.dumps(datadict);
	return json_out;
	#except:
		#pass

t1 = Thread(target=scavenge_data_q, args=(q,))
t2 = Thread(target=update_lcd)
t1.start()
t2.start()

#Server definitions
@route('/data.json')
def process():
	response.content_type = 'application/json; charset=UTF8'
	response.add_header("Cache-Control", "no-cache")
	return get_json()						#Send JSON to connected clients

@route('/<filename>')
def server_static(filename):
	print(filename)
	if filename == "restart.html":
		print("Restart Trig'd")
		subprocess.Popen('sleep 1; systemctl restart server.service', shell=True)
	response = static_file(filename, root="Server")
	response.set_header('Content-Language', 'de')
	response.add_header("Cache-Control", "no-cache")
	return response;						#Send requested file from Server root directory
@route('/de/<filename>')
def server_static(filename):
	print(filename)
	if filename == "restart.html":
		print("Restart Trig'd")
		subprocess.Popen('sleep 1; systemctl restart server.service', shell=True)
	response = static_file(filename, root="Server/de/")
	response.set_header('Content-Language', 'de')
	response.add_header("Cache-Control", "no-cache")
	return response;						#Send requested file from Server root directory
@route('/')
def index():
	response = static_file("index.html", root="Server")
	response.set_header('Content-Language', 'de')
	response.add_header("Cache-Control", "no-cache")
	return response;						#Send the index.html as default for accessing the server

#@error(404) 
#def error404(error):
#	response = static_file("404.html", root="Server")
#	return response;


run(host='0.0.0.0', port=80, debug=True)										#Start the Server and listen on all interfaces and port 80
